### PR TITLE
[remat] Change remat lowering to XLA::Conditional

### PR DESCRIPTION
**Pull request by trevorcai**
`jax.remat` creates rematerializing passes that don't have data dependencies on
the actual loss-computing forward pass. This means that the XLA scheduler was
free to schedule the remat forward pass before the loss-computing pass,
defeating the goal of saving accelerator memory with `jax.remat`.

In practice, it sometimes did for my workloads.

This change expresses the lowering of remat_call(f) as:
Conditional(true, inputs, f, inputs, dummy_f).

In the common case of `jax.grad(jax.remat(f))`, the content of the
lowered remat_call are both the forwards & backwards; that is, the
incoming cotangents are part of the args.

Additionally, Conditional (AFAIK) is un-inlineable in the sense that it
doesn't execute until all its inputs (e.g. cotangents!) are available.

Downsides:

- AFAICT, we can no longer interleave computation in/outside the
  rematerialized block.
- Potentially, lower performance. I do not observe this in my tests.

### CODE REVIEW

**Review by jekbradbury — APPROVED** [review]

**Comment by mattjj** [comment]
This looks good, but IIUC @trevorcai suggested we wait to merge until he runs more tests. Let us know!

**Comment by trevorcai** [comment]
Had to include one-line change to work around an upstream XLA bug which got the parameter replication on the Conditionals wrong somehow. I think this is ready to merge.

**Review by mattjj — APPROVED** [review]
LGTM, thanks!
